lib/PhasarLLVM/TypeHierarchy/DIBasedTypeHierarchy.cpp

```diff
@@ -0,0 +1,315 @@
+    SubTypes.insert(VertexTypes[Index]);
+  }
+
+  return SubTypes;
+}
+
+[[nodiscard]] bool DIBasedTypeHierarchy::isSuperType(ClassType Type,
+                                                     ClassType SuperType) {
+  return isSubType(SuperType, Type); // NOLINT
+}
+
+[[nodiscard]] auto DIBasedTypeHierarchy::getSuperTypes(ClassType Type)
+    -> std::set<ClassType> {
+  // TODO: implement (low priority)
+  llvm::report_fatal_error("Not implemented");
+}
+
+[[nodiscard]] bool DIBasedTypeHierarchy::hasVFTable(ClassType Type) const {
+  const auto TypeIndex = TypeToVertex.find(Type);
+
+  if (TypeIndex == TypeToVertex.end() || TypeIndex->second >= VTables.size()) {
+    return false;
+  }
+
+  return !VTables[TypeIndex->second].empty();
+}
+
+[[nodiscard]] auto DIBasedTypeHierarchy::getVFTable(ClassType Type) const
+    -> const VFTable<f_t> * {
+  const auto TypeIndex = TypeToVertex.find(Type);
+  assert(TypeIndex != TypeToVertex.end());
+  return &VTables[TypeIndex->getSecond()];
+}
+
+void DIBasedTypeHierarchy::print(llvm::raw_ostream &OS) const {
+  size_t VertexIndex = 0;
+  size_t EdgeIndex = 0;
+
+  OS << "Type Hierarchy\n";
+  for (const auto &CurrentVertex : TransitiveClosure) {
+    OS << VertexTypes[VertexIndex]->getName() << "\n";
+    for (size_t I = 0; I < TransitiveClosure.size(); I++) {
+      if (EdgeIndex != VertexIndex && CurrentVertex[I]) {
+        OS << "--> " << VertexTypes[EdgeIndex]->getName() << "\n";
+      }
+      EdgeIndex++;
+    }
+    VertexIndex++;
+    EdgeIndex = 0;
+  }
+
+  OS << "VFTables:\n";
+
+  size_t TypeIndex = 0;
+  for (const auto &Type : VertexTypes) {
+    const auto &VTable = VTables[TypeIndex];
+
+    OS << Type->getName() << ": ";
+
+    // get all function names for the llvm::interleaveComma function
+    llvm::SmallVector<std::string, 6> Names;
+    for (const auto &Function : VTable.getAllFunctions()) {
+      if (Function) {
+        Names.push_back(Function->getName().str());
+      } else {
+        Names.push_back("<none>");
+      }
+    }
+    // prints out all function names, seperated by comma, without a trailing
+    // comma
+    llvm::interleaveComma(Names, OS);
+    OS << "\n";
+    TypeIndex++;
+  }
+
+  OS << "\n";
+}
+
+[[nodiscard]] nlohmann::json DIBasedTypeHierarchy::getAsJson() const {
+  /// TODO: implement
+  llvm::report_fatal_error("Not implemented");
+}
+
+void DIBasedTypeHierarchy::printAsDot(llvm::raw_ostream &OS) const {
+  OS << "digraph TypeHierarchy{\n";
+
+  if (TransitiveClosure.size() != VertexTypes.size()) {
+    llvm::errs() << "TC.size(): " << TransitiveClosure.size()
+                 << " VT.size(): " << VertexTypes.size();
+    llvm::report_fatal_error(
+        "TransitiveClosure and VertexType size not equal.");
+    return;
+  }
+
+  // add all edges
+  size_t CurrentRowIndex = 0;
+  for (const auto &Row : TransitiveClosure) {
+    for (const auto &Cell : Row.set_bits()) {
+      if (Row[Cell]) {
```

Comment:
actually this is always true. set_bits() lets you iterate over all the positions where 1 is set.